### PR TITLE
eval: Make writefile() able to disable fsync()

### PR DESCRIPTION
**Pull request by ZyX-I**
Ref #6420

### CODE REVIEW

**Comment by ZyX-I** [comment]
As fsync() is a safer option, it is on by default. As other flags enable something and are lowercase, flag that disables fsync() is uppercase.

**Comment by justinmk** [comment]
Shouldn't writefile() default behavior be decided by 'fsync' option? And then overridden with s/S flag pair.

**Comment by blueyed** [comment]
@justinmk 
Good idea, but the documentation for `'fsync'` could be interpreted as not applying here (but only to `:w` etc).
I'm assuming that Vim does not `fsync` with `writefile()`, does it?

**Comment by blueyed** [comment]
Just for reference: any flag for `writefile` will trigger the `binary` flag in old Vim versions, where a trueish value enables this, and no other flags were known yet.  So using `S` would always trigger it, but that should be handled in plugins by checking for a patch level probably.

**Comment by justinmk** [comment]
> Good idea, but the documentation for 'fsync' could be interpreted as not applying here (but only to :w etc).

Right, but I'm suggesting that we apply it anyway.

**Comment by ZyX-I** [comment]
@blueyed Vim does not `fsync()` neither with `writefile()` nor when writing viminfo (second user of buffered writing is ShaDa). Basically you can’t use `FILE*`, `fsync` and keep up with the standard: even if `fileno()` is available (it is POSIX, not C99), it is required that you should not mix unistd API with stdio API, this may yield unexpected results.

Though `fsync()` with the whole fd-related API is also BSD/POSIX/SUS and not C99, yet available on Windows (but specifically `fsync()` is absent, libuv uses `FlushFileBuffers` and we are using libuv).

**Comment by ZyX-I** [comment]
`ag` finds exactly three occurrences of `p_fs` (fsync option): fileio.c with `#ifdef UNIX` guard (BTW, I think I should remove it, libuv has windows implementation for that), globals.h and options.lua.

Note that our help does not warn about “systems without fsync implementation” like Vim documentation. Yet there is a guard around the only place where `p_fs` is checked.

**Comment by justinmk** [comment]
> ag finds exactly three occurrences of p_fs (fsync option): fileio.c with #ifdef UNIX guard (BTW, I think I should remove it, libuv has windows implementation for that), 

Yes, probably should remove it. Most likely we overlooked it because Windows was "todo" for a long time. But the documentation was updated in anticipation of convergence.

**Comment by ZyX-I** [comment]
My watchdog is working (which restarts single includes test), but this is not very useful: “hang up” was replaced with “CMakeFiles/test-includes-api-window-h.dir/auto/api/window.h.test-include.c.o: file not recognized: File truncated” and “/usr/bin/ld: cannot find CMakeFiles/test-includes-mark_defs-h.dir/auto/mark_defs.h.test-include.c.o: No such file or directory”.

File: test/functional/ui/cursor_spec.lua [hunk 1]
```diff
@@ -1,8 +1,7 @@
 local helpers = require('test.functional.helpers')(after_each)
 local Screen = require('test.functional.ui.screen')
-local clear, feed, meths = helpers.clear, helpers.feed, helpers.meths
-local insert, execute = helpers.insert, helpers.execute
-local eq, funcs = helpers.eq, helpers.funcs
+local clear, meths = helpers.clear, helpers.meths
+local eq = helpers.eq
```
Comment (ZyX-I) [inline test/functional/ui/cursor_spec.lua]: BTW, how that appeared in master? I did not touch the file, yet tests fail.

**Comment by ZyX-I** [comment]
Will have to merge despite linter error: I am not really up to splitting `buf_write`.

File: test/functional/ui/cursor_spec.lua (hunk 1 above)
Comment (justinmk) [inline test/functional/ui/cursor_spec.lua]: my fault

**Review by justinmk — APPROVED** [review]

**Comment by blueyed** [comment]
I wonder if it should really use fsync by default?!
Typically you call this manually, or implicit when closing a file.
(A file cannot be closed though in VimL, can it?)

**Comment by ZyX-I** [comment]
@blueyed Typically software does not bother to call `fsync()`, it is never done implicitly and “manually” you can only fsync all files in the whole system. What does call fsync() is something old-school like Vim: software whose developers assume that crashing system is relatively common. E.g. in Python you will get `fsync()` only if you use `posix` module, neither `file.flush()` nor `file.close()` do that.

**Comment by ZyX-I** [comment]
@blueyed And `writefile()` does close file. As well as e.g. `:write` first opens it and then closes back (after writing).

**Comment by ZyX-I** [comment]
When file is opened in a buffer it is not opened from the system POV: file descriptor is created and associated with the file only temporary, when you read or write a file, disappearing after finishing reading or writing.

**Comment by justinmk** [comment]
> I wonder if it should really use fsync by default?!

You can control the default, e.g. `:set nofsync`.

**Pull request merged**